src/parser.ts

```diff
@@ -0,0 +1,261 @@
+// [VexFlow](http://vexflow.com) - Copyright (c) Mohit Muthanna 2010.
+//
+// A generic text parsing class for VexFlow.
+
+import { Vex } from './vex';
+import { Grammar } from './easyscore';
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// TODO: Move to easyscore.ts
+type RuleFunction = () => Rule;
+type TriggerFunction = (state?: Result) => void;
+interface Rule {
+  // Lexer Rules
+  token?: string; // The token property is a string that is compiled into a RegExp.
+  noSpace?: boolean; // TODO: None of the EasyScore rules specify noSpace, so it is not used anywhere.
+
+  // Parser Rules
+  expect?: RuleFunction[];
+  zeroOrMore?: boolean;
+  oneOrMore?: boolean;
+  maybe?: boolean;
+  or?: boolean;
+  run?: TriggerFunction;
+}
+// END: Move to easyscore.ts
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+// To enable logging for this class. Set `Vex.Flow.Parser.DEBUG` to `true`.
+function L(...args: any[]) {
+  if (Parser.DEBUG) Vex.L('Vex.Flow.Parser', args);
+}
+
+export const X = Vex.MakeException('ParserError');
+
+const NO_ERROR_POS = -1;
+
+interface Result {
+  success: boolean;
+
+  // Lexer Results
+  pos?: number;
+  incrementPos?: number;
+  matchedString?: string;
+
+  // Parser Results
+  matches?: Match[];
+  numMatches?: number;
+  results?: Result[];
+  errorPos?: number; // Set to NO_ERROR if successful. N if there is an error in the string.
+}
+
+// Converts parser results into an easy to reference list that can be used in triggers.
+// eventually returns: string (including empty strings) | nested array | null
+type Match = null | string | Array<Match>;
+function flattenMatches(r: Result | Result[]): Match {
+  if ('matchedString' in r) return r.matchedString as string; // string
+  if ('results' in r) return flattenMatches(r.results as Result[]);
+  const results = r as Result[];
+  if (results.length === 1) return flattenMatches(results[0]);
+  if (results.length === 0) return null; // null
+  return results.map(flattenMatches); // nested array
+}
+
+// This is the base parser class. Given an arbitrary context-free grammar, it
+// can parse any line and execute code when specific rules are met (e.g.,
+// when a string is terminated.)
+export class Parser {
+  public static DEBUG: boolean = false;
+
+  private grammar: Grammar;
+  private line: string;
+  private pos: number;
+  private errorPos: number;
```

Comment:
Please change to `protected` we are using that in general. `private` only if we want to force the child to use access functions making some extra required processing.

Reply:
Sounds good. Will do.

I actually like when libraries minimize use of `private` methods and instance variables because it allows a developer to subclass and override behaviors when necessary.

I initially chose `private` in this case, because the variables like line, pos, and errorPos are used mostly like temporary internal state for the Parser. They shouldn't really be accessed by anyone outside, because their state continually changes as we parse the grammar.

If the Parser were built in a more functional / closure-heavy fashion, these variables might have been hidden local vars that were captured by the closure.

Also, during migration, setting everything to private has a nice side effect. The typescript compiler errors will tell me which external classes are trying to access the class I'm migrating! :-)

Reply:
You are right this is nice :)